### DISCUSSION TITLE
Controlling the CF with an asynchronous version of the motion commander

### DISCUSSION THREAD

**Discussion by EJKT** [bitcraze/discussions#134] [Q&A]
I  apologize for the question I posted yesterday or the day before to which I could have easily found the answer in the documentation. There is however still something that is not clear to me: can you control the motion commander asynchronously? This would be useful for me because it would allow making a function that for example takes x and y displacement from the current position as an input and then makes the drone do that. If one were to use the synchronous way of doing things, it would land after this function which is not desired. Maybe there are alternatives, just let me know.

Thanks in advance,

Ethan Thiebaut

**Comment by knmcguire** [comment]
Hi!

There is an step-by-step example that at least has async logging, but not that for the motion commander I'm afraid: https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/user-guides/sbs_motion_commander/. 

I'm sure that if you start the [motion commander instance](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/api/cflib/positioning/motion_commander/) in the right way, and find a way to pass that instance to the call back functions, this should be possible. However there is a lot of logic going on in that class that might block it from flying if the connection is setup probably, since it is meant as a wrapper around the basic commander class to make things (supposably) simpler. So if you want to do something special like that maybe the motion commander is not the right place to look. 

Their might be an 'easier' way How about you use the [commander class](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/api/cflib/crazyflie/commander/) instance? This is one step down in high-level-ness, but you are able to more easily send setpoints to the crazyflie (like send_hover_setpoint() ). As it is low level, it needs to be ensured that the setpoint is send constantly with a half-second interval at least. 

There are some examples in the python library repo that you could look at that use the basic commander class:
- [examples/positioning/flowsequenceSync.py](https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/positioning/flowsequenceSync.py)
- [examples/multiranger/multiranger_pointcloud.py](https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/multiranger/multiranger_pointcloud.py) (this used the send setpoints in async way as well)

**Reply by EJKT** [reply]
Hi, thanks for the response. The commander class seems to be a good replacement. I am wondering if there exists a way to set the velocity at which at the set_position_setpoint command is executed, just with a flow deck? For example if I want to reach (x,y,z,yaw) in a given time the only way to do that seems to be through computing the roll, pitch, yaw and thrust values and how long they should be maintained. I saw that there is a send_velocity_world_setpoint function, but is this meant for reference frames generated by the flowdeck (as this is relative and not "world"). Anyways please let me know what your solution would be.
Thanks.

**Reply by knmcguire** [reply]
Ah I think that probably the function [ 'send_hover_setpoint()'](https://github.com/bitcraze/crazyflie-lib-python/blob/03053503be927d8e259c1d0eeb5127a9fbb8c318/cflib/crazyflie/commander.py#L116) is the one you want! Vx and vy are in body coordinate, yaw is in rate (velocity) and the height is absolute. That is what we use for flowdeck measurements.

I see that the explanation of the function is not correct so I'll update that right away.

**Reply by EJKT** [reply]
Yes, indeed I saw that function but it does not allow you to set a velocity in the z direction. How would one do that? It cannot exactly be done with regular delays as reaching the given zdistance will be done at a preset velocity.

**Reply by knmcguire** [reply]
Ah no it does not.  You will need to write your own function for that. The problem is that there are preset setpoint types that correlate with the [firmware generic commander types](https://github.com/bitcraze/crazyflie-firmware/blob/79aead2deafe0f9c5134422566c2f80e2db96613/src/modules/src/crtp_commander_generic.c#L65), so that will require a firmware change as well. 

Another option is to use the [send_setpoint ](https://github.com/bitcraze/crazyflie-lib-python/blob/03053503be927d8e259c1d0eeb5127a9fbb8c318/cflib/crazyflie/commander.py#L63) directly, and change parameters in the [flightmode parameter group ](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#flightmode), but again there is nowhere within the [crtp_commander_rpyt.c](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/crtp_commander_rpyt.c) that the velocity_body is set to true. 

Or you managed to write something that just adds the increment on top of the estimated zdistance at every time that you send the setpoint with the absolute distance, which is what the [pointcloud script](https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/multiranger/multiranger_pointcloud.py) is doing. It is not ideal but it will prevent you from changing the firmware.

**Comment by knmcguire** [comment]
The high level commander you could also have a look at. As this is for more onboard autonomy you don't need to send it setpoints every 0.5 seconds, but if you lose connection it won't turn off the motors either.

Here is the api for that: https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/api/cflib/crazyflie/high_level_commander/ and it has a go_to functionality that can be done relative of your current position.

**Reply by EJKT** [reply]
Ok, then it seems that the HL commander is the best way to go. I have checked it out and have seen that in addition to the go_to function there are others for landing, taking off, following trajectories and more. Are these are all made for the flow deck? And what kind of set up do I need in my code for the HL commander? I assume it is different from the commander and motion commander classes.

**Reply by knmcguire** [reply]
technically they were made for mocap, but they should work for any positioning system, including the flowdeck. 

This is an example that uses the hl commander: https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/autonomy/autonomous_sequence_high_level.py. Just comment out [the trajectory parts](https://github.com/bitcraze/crazyflie-lib-python/blob/b0366efe9e9a948df42edd4dddad97883194b028/examples/autonomy/autonomous_sequence_high_level.py#L175) because that requires quite some space.

I've given a lecture to an EPFL class a while ago, where I explained about the types of commanders as well, so perhaps that can make it more clear: https://www.youtube.com/watch?v=6s8i-nhPjt0&list=PLj9XMmQVSr-C-ZXeWWyZgeZ7C4g6A5nPj&index=2&t=2696s
